File: utils/graph.cpp
#include <cstdint>
#include <cassert>
#include <iostream>
#include <list>
#include <map>
#include <string>


typedef ::std::string StdString;
// c++11
// template <typename T> using StdList = ::std::list<T>;
// template <typename KEY, typename VALUE>
    // using StdMap = ::std::map<KEY, VALUE>;
template <typename T>
class StdList : public ::std::list<T>
{};
template <typename KEY, typename VALUE>
class StdMap : public ::std::map<KEY, VALUE>
{};

namespace uchiha
{
    namespace ryuuzaki
    {
        class Graph;
    }
}

class uchiha::ryuuzaki::Graph
{
private:
    class Edge
    {
    public:
        explicit Edge(void) : __weight__(0)
        {}
        explicit Edge(intptr_t weight) : __weight__(weight)
        {}

    public:
        intptr_t GetWeight(void) const
        {
            return __weight__;
        }

    private:
        intptr_t __weight__;
    };

    typedef StdMap<StdString, Edge> VertexOutMap;

    class Vertex
    {
    public:
        void AddEdge(StdString const &name, Edge &edge)
        {
            __out_map__[name] = edge;

            return;
        }

        intptr_t FindEdge(StdString const &name)
        {
            return (__out_map__.end() != __out_map__.find(name));
        }

        VertexOutMap const &GetOutMap(void) const
        {
            return __out_map__;
        }

        intptr_t EdgeCount(void) const
        {
            return __out_map__.size();
        }

        intptr_t DelEdge(StdString const &name)
        {
            if (__out_map__.end() == __out_map__.find(name)) {
                return -1;
            }

            __out_map__.erase(name);

            return 0;
        }

        void Print(void) const
        {
            for (StdMap<StdString, Edge>::const_iterator it = __out_map__.begin();
                 __out_map__.end() != it;
                 ++it)
            {
                ::std::cout << it->first
                            << "(" << it->second.GetWeight() << "),";
            }
        }

    private:
        VertexOutMap __out_map__;
    };

public:
    explicit Graph(void);
    intptr_t Init();
    ~Graph(void); // inheritance denied

public:
    intptr_t VertexCount(void) const
    {
        return __vertexes__.size();
    }

    intptr_t EdgeCount(void) const
    {
        return __edge_count__;
    }

    void AddVertex(StdString name);
    intptr_t DelVertex(StdString name);
    void AddEdge(StdString from, StdString to, intptr_t weight = 0);
    intptr_t DelEdge(StdString from, StdString to);
    void Print(void) const;

private:
    typedef StdMap<StdString, Vertex> VertexMap;
    typedef StdMap<StdString, intptr_t> VertexIndegreeMap;

    intptr_t __edge_count__;
    VertexMap __vertexes__;
    VertexIndegreeMap __indegree_of_vertexes__;
};

::uchiha::ryuuzaki::Graph::Graph(void)
    : __edge_count__(0)
{}

intptr_t ::uchiha::ryuuzaki::Graph::Init(void)
{
    return 0;
}

::uchiha::ryuuzaki::Graph::~Graph(void)
{}

void ::uchiha::ryuuzaki::Graph::AddVertex(StdString name)
{
    __vertexes__[name];
    __indegree_of_vertexes__[name];

    return;
}

intptr_t ::uchiha::ryuuzaki::Graph::DelVertex(StdString name)
{
    if (__vertexes__.end() == __vertexes__.find(name)) {
        return -1;
    }

    for (VertexMap::iterator it = __vertexes__.begin();
         __vertexes__.end() != it;
         ++it)
    {
        if (0 == it->second.DelEdge(name)) {
            --__edge_count__;
        }
    }
    __edge_count__ -= __vertexes__[name].EdgeCount();

    VertexOutMap const &out_map = __vertexes__[name].GetOutMap();
    for (VertexOutMap::const_iterator it = out_map.begin();
         out_map.end() != it;
         ++it)
    {
        --__indegree_of_vertexes__[it->first];
    }

    __indegree_of_vertexes__.erase(name);
    __vertexes__.erase(name);

    return 0;
}

void ::uchiha::ryuuzaki::Graph::AddEdge(StdString from,
                                        StdString to,
                                        intptr_t weight)
{
    Edge e(weight);

    AddVertex(from);
    AddVertex(to);

    __vertexes__[from].AddEdge(to, e);
    ++__indegree_of_vertexes__[to];

    ++__edge_count__;

    return;
}

intptr_t ::uchiha::ryuuzaki::Graph::DelEdge(StdString from, StdString to)
{
    if (__vertexes__.end() == __vertexes__.find(from)) {
        return -1;
    }

    if (-1 == __vertexes__[from].DelEdge(to)) {
        return -1;
    }

    --__indegree_of_vertexes__[to];

    --__edge_count__;

    return 0;
}

void ::uchiha::ryuuzaki::Graph::Print(void) const
{
    ::std::cout << "total vertexes: " << VertexCount()  << ::std::endl;
    ::std::cout << "total edges: " << EdgeCount()  << ::std::endl;
    for (VertexMap::const_iterator it = __vertexes__.begin();
         __vertexes__.end() != it;
         ++it)
    {
        ::std::cout << it->first
                    << "(" << __indegree_of_vertexes__.at(it->first) << "): ";
        it->second.Print();
        ::std::cout << ::std::endl;
    }

    return;
}

int main(int argc, char *argv[])
{
    ::uchiha::ryuuzaki::Graph g;

#define TEST 0
#if 0 == TEST

    g.AddEdge("home", "school", 12);
    g.AddEdge("home", "company");
    g.AddEdge("guangzhou", "shenzhen", 100);
    g.DelVertex("home");
    /*assert(0 == g.DelEdge("guangzhou", "shenzhen"));
    assert(0 == g.DelEdge("home", "school"));
    assert(0 == g.DelEdge("home", "company"));
    assert(-1 == g.DelEdge("home", "home"));
    assert(-1 == g.DelEdge("shenzhen", "guangzhou"));
    assert(-1 == g.DelEdge("guangzhou", "shenzhen"));*/
    g.Print();

#elif 1 == TEST

    g.AddEdge("v1", "v2");
    g.AddEdge("v1", "v3");
    g.AddEdge("v1", "v4");
    g.AddEdge("v2", "v4");
    g.AddEdge("v2", "v5");
    g.AddEdge("v3", "v6");
    g.AddEdge("v4", "v3");
    g.AddEdge("v4", "v6");
    g.AddEdge("v4", "v7");
    g.AddEdge("v5", "v4");
    g.AddEdge("v5", "v7");
    g.AddEdge("v7", "v6");

    g.Print();

#endif

    return 0;
}

File: utils/graph.py
#! /usr/bin/env python
# -*- coding:utf-8 -*-


import sys
import copy


class graph:
    ''' vertex '''
    class vertex:
        ''' edge '''
        class edge:
            def __init__(self, weight):
                self.__weight = weight;

            def get_weight(self):
                return self.__weight

        def __init__(self):
            self.__edges = {}

        def add_edge(self, to, weight):
            if not self.exist_edge(to):
                self.__edges[to] =  graph.vertex.edge(weight)

        def get_edges(self):
            return self.__edges

        def exist_edge(self, to):
            return to in self.__edges.keys()

        def del_edge(self, to):
            if self.exist_edge(to):
                del self.__edges[to]

        def printme(self):
            for (k, v) in self.__edges.items():
                sys.stdout.write("%s(%d)|" % (k, v.get_weight()))
            sys.stdout.write("\n")

    ''' graph '''
    def __init__(self):
        self.__vertexes = {}
        self.__vertexes_indegree = {}

    def __add_vertex(self, name):
        self.__vertexes[name] = graph.vertex()
        self.__vertexes_indegree[name] = 0

    def add_vertex(self, name):
        if name not in self.__vertexes.keys():
            self.__add_vertex(name)

    def del_vertex(self, name):
        if name not in self.__vertexes.keys():
            return -1

        ''' 删除所有到该顶点的边 '''
        for (k, v) in self.__vertexes.items():
            v.del_edge(name)

        ''' 降低所有后继顶点的入度 '''
        for (k, v) in self.__vertexes[name].get_edges().items():
            self.__vertexes_indegree[k] -= 1

        ''' 删除顶点极其出边 '''
        del self.__vertexes[name]
        del self.__vertexes_indegree[name]

        return 0

    def add_edge(self, frm, to, weight = 1):
        if frm not in self.__vertexes.keys():
            self.__add_vertex(frm)
        if to not in self.__vertexes.keys():
            self.__add_vertex(to)

        if not self.__vertexes[frm].exist_edge(to):
            self.__vertexes[frm].add_edge(to, weight)
            self.__vertexes_indegree[to] += 1

    def del_edge(self, frm, to):
        if frm not in self.__vertexes.keys():
            return -1

        if to not in self.__vertexes.keys():
            return -1

        if not self.__vertexes[frm].exist_edge(to):
            return -1

        self.__vertexes[frm].del_edge(to)
        self.__vertexes_indegree[to] -= 1

        return 0

    ''' 拓扑排序 '''
    def topsort(self):
        rslt = []
        g = copy.deepcopy(self)

        while g.__vertexes_indegree:
            size_before = len(g.__vertexes_indegree)
            for k in g.__vertexes_indegree.keys():
                if 0 == g.__vertexes_indegree[k]:
                    rslt.append(k)
                    g.del_vertex(k)
            size_after = len(g.__vertexes_indegree)

            if size_before == size_after:
                return []

        return rslt

    ''' dijkstra '''
    def dijkstra(self, name):
        if name not in self.__vertexes.keys():
            return []

    def printme(self):
        print self.__vertexes
        print self.__vertexes_indegree
        for (k, v) in self.__vertexes.items():
            sys.stdout.write("%s(%d): " % (k, self.__vertexes_indegree[k]))
            v.printme()


if "__main__" == __name__:
    g = graph()

    ''' top sort
    g.add_edge("v1", "v2")
    g.add_edge("v1", "v3")
    g.add_edge("v1", "v4")
    g.add_edge("v2", "v4")
    g.add_edge("v2", "v5")
    g.add_edge("v3", "v6")
    g.add_edge("v4", "v3")
    g.add_edge("v4", "v6")
    g.add_edge("v4", "v7")
    g.add_edge("v5", "v4")
    g.add_edge("v5", "v7")
    g.add_edge("v7", "v6")
    print g.topsort()
    g.printme()
    '''

    ''' dijkstra '''
    g.add_edge("A", "B", 20)
    g.add_edge("A", "D", 80)
    g.add_edge("A", "G", 90)
    g.add_edge("B", "F", 10)
    g.add_edge("F", "C", 10)
    g.add_edge("F", "D", 40)
    g.add_edge("C", "D", 10)
    g.add_edge("C", "F", 50)
    g.add_edge("C", "H", 20)
    g.add_edge("D", "C", 10)
    g.add_edge("D", "G", 20)
    g.add_edge("G", "A", 20)
    g.add_edge("E", "G", 30)
    g.add_edge("E", "B", 50)
    print g.dijkstra("Z")

    sys.exit(0)
